### DISCUSSION TITLE
Clients IPv4+IPv6 work only with IPv4+IPv6 servers and also about other varaints

### DISCUSSION THREAD

**Discussion by Perlover** [XTLS/Xray-core#2487] [General]
Hello everyone!

Recently I ran into a problem about which I have not seen a description anywhere. Perhaps my experience will be useful to someone else.

I was already sitting down to write about how the VLESS + REALITY protocol is blocked in Dubai at the level of the "Du" mobile operator. And now, almost before sending this ticket, I discovered the real causes of the problem.

Imagine a configuration that many people can have. The server is IPv4 only and does not have a working IPv6.

Clients: v2rayNG (Android) or Nekoray (computer, Linux). But they are connected to a provider that gives both IPv4 and IPv6 (in my case, it was the operator "Du" and my phone as an access point for these devices: phone and computer).

Both connected devices, when VLESS + REALITY is turned on, can work for a few seconds and then stop. The Speedtest utility shows that there is no connection. Almost all sites do not open. It looks like Du has learned to block the REALITY raid using DPI.

But only by chance, when I still managed to open the site sometimes whatismyipaddress.com, I figured out what the reason is. The site showed me the IPv4 address as the address of my xray server, and the IPv6 showed the one that was given to me by the Du's DHCP provider for phone.

It turns out that all my devices (computer with TUN mode or connected phones to the access point, and the phone itself in the SIM card Du) - they all support IPv6, and when I worked with them, they, where possible, started using IPv6 addresses and they went to the tunnel VLESS+REALITY on the IPv4 server side. And he, of course, did not know what to do with them and dumped them.

I tried to solve the problem so that I deployed xray server on another server where there is IPv4 + IPv6, and the first IPv4 server had outbounds on VLESS+TCP with `"ip": ["::/0"]` rules. But I have not been able to set up such a configuration reliably. The channel worked, but poorly (apparently this is due to DNS and incomplete IPv6 support on the first server).

As a result, I abandoned the idea of wrapping IPv6 traffic to another node and created a new configuration on an IPv4 + IPv6 server. After I checked, all clients began to work reliably with such a server.

From this I concluded:

If you have an **IPv4 xray server only**, then it **will only work with those clients that are connected to an IPv4 only provider**. The **IPv4 + IPv6 server** will work **with both IPv4-only clients and IPv4+IPv6 clients**.

**Comment by testcaoy7** [comment]
My cellular ISP is China Mobile and they do use dual stack network.
Never encounter any problem.

**Reply by Perlover** [reply]
Do you have only a xray IPv4 only server? And do you use this IPv4 only xray server from this cellular ISP?

**Reply by testcaoy7** [reply]
I do have ipv4-only VPS. If I may ask, do you use dual-stack DNS records for your domain?

**Reply by Perlover** [reply]
Explain your question in more detail. I didn't understand what you call "your domain". XRay server? It does not have a domain - only IPv4 is specified in the client configuration. The domain of the site that the server pretends to be? But this is not my domain and I don't see any connection with the domain here.

> I do have ipv4-only VPS.

I wanted to clarify once again - does your provider give you IPv4 and specifically IPv6? Which IPv6 shows, for example, the site https://whatismyipaddress.com / when are you connected from your XRay server? I was shown IPv4 of my server, and IPv6 was about the Dubai mobile operator.

**Comment by us254** [comment]
routes IPv6 traffic to the blackhole outbound, effectively blocking it
 
 
 ```
 {
  "outbounds": [
    {
      "protocol": "freedom",
      "settings": {}
    },
    {
      "tag": "blocked",
      "protocol": "blackhole",
      "settings": {}
    }
  ],
  "routing": {
    "rules": [
      {
        "type": "field",
        "ip": [
          "::/0"
        ],
        "outboundTag": "blocked"
      }
    ]
  }
}
 ```
